cli/replicaset/cmd/upgrade.go

```diff
@@ -0,0 +1,256 @@
+	run  running.InstanceCtx
+	conn connector.Connector
+}
+
+//go:embed lua/upgrade.lua
+var upgradeMasterLua string
+
+type syncInfo struct {
+	LSN uint64  `mapstructure:"lsn"`
+	IID uint32  `mapstructure:"iid"`
+	Err *string `mapstructure:"err"`
+}
+
+// filterReplicasetsByAliases filters the given replicaset list by chosen aliases and
+// returns the chosen replicasets. If a non-existent alias is found, it returns an error.
+func filterReplicasetsByAliases(replicasets replicaset.Replicasets,
+	chosenReplicasetAliases []string) ([]replicaset.Replicaset, error) {
+	// If no aliases are provided, return all replicasets.
+	if len(chosenReplicasetAliases) == 0 {
+		return replicasets.Replicasets, nil
+	}
+
+	// Create a map for fast lookup of replicasets by alias.
+	replicasetMap := make(map[string]replicaset.Replicaset)
+	for _, rs := range replicasets.Replicasets {
+		replicasetMap[rs.Alias] = rs
+	}
+
+	var chosenReplicasets []replicaset.Replicaset
+	for _, alias := range chosenReplicasetAliases {
+		rs, exists := replicasetMap[alias]
+		if !exists {
+			return nil, fmt.Errorf("replicaset with alias %q doesn't exist", alias)
+		}
+		chosenReplicasets = append(chosenReplicasets, rs)
+	}
+
+	return chosenReplicasets, nil
+}
+
+func Upgrade(discoveryCtx DiscoveryCtx, opts UpgradeOpts) error {
+	replicasets, err := getReplicasets(discoveryCtx)
+	if err != nil {
+		return err
+	}
+
+	replicasets = fillAliases(replicasets)
+	replicasetsToUpgrade, err := filterReplicasetsByAliases(replicasets,
+		opts.ChosenReplicasetAliases)
+	if err != nil {
+		return err
+	}
+
+	return internalUpgrade(replicasetsToUpgrade, opts.LsnTimeout)
+}
+
+func internalUpgrade(replicasets []replicaset.Replicaset, lsnTimeout int) error {
+	for _, replicaset := range replicasets {
+		err := upgradeReplicaset(replicaset, lsnTimeout)
+		if err != nil {
+			fmt.Printf("• %s: error\n", replicaset.Alias)
+			return fmt.Errorf("replicaset %s: %w", replicaset.Alias, err)
+		}
+		fmt.Printf("• %s: ok\n", replicaset.Alias)
+	}
+	return nil
+}
+
+func getInstanceConnector(instance replicaset.Instance) (connector.Connector, error) {
+	run := instance.InstanceCtx
+	fullInstanceName := running.GetAppInstanceName(run)
+	if fullInstanceName == "" {
+		fullInstanceName = instance.Alias
+	}
+	if fullInstanceName == "" {
+		fullInstanceName = "unknown"
+	}
+
+	// Try to connect via unix socket.
+	conn, err := connector.Connect(connector.ConnectOpts{
+		Network: "unix",
+		Address: run.ConsoleSocket,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("instance %s failed to connect via UNIX socket "+
+			": %w", fullInstanceName, err)
+	}
+	return conn, nil
+}
+
+func collectRWROInfo(replset replicaset.Replicaset) (*instanceMeta, []instanceMeta,
+	error) {
+	var master *instanceMeta = nil
+	var replicas []instanceMeta
+	for _, instance := range replset.Instances {
+		run := instance.InstanceCtx
+		fullInstanceName := running.GetAppInstanceName(run)
+		conn, err := getInstanceConnector(instance)
```

Comment:
Im still not sure about this. What if some connections were created, but after that we fail on some of the checks below:
```
if instance.Mode == replicaset.ModeUnknow
```
or
```
if isRW && master != nil
```
or even after an error in `upgradeMaster`, `waitLSN` or `snapshot`.
Than error is returned, but none of the created connection are closed. Is it possible to achieve this?